Make this crate no_std

linkmauve opened the pull request:
As nothing actually depends on `std`, there is no need to keep a dependency on it.  This allows users such as the [jid](https://docs.rs/jid) crate to actually be `no_std` as well.

linkmauve committed:
Make this crate no_std

This allows users such as the jid crate to actually be no_std as well.

core::error::Error got stabilized back in Rust 1.81.0, but in order to
support older Rust versions we added a std feature.

linkmauve commented:
Ah oops, that works in nightly but doesn’t yet in stable, due to `core::error::Error` not being exported yet… I could add a `std` feature that would be default and impl this trait, [just like in jid](https://gitlab.com/xmpp-rs/xmpp-rs/-/commit/48f77acb690c025952f385a58f7850e5c63401e4), or just keep that PR open until that export gets stabilised, which option do you prefer?

sfackler commented:
I think a `std` feature seems reasonable. If it becomes a no-op in the future once Error-in-core stabilizes that's fine.

linkmauve commented:
There, I have added an optional but enabled by default `std` feature.

sfackler commented:
Can you also update the CI logic to also test with the std feature disabled?

LGTM otherwise

linkmauve commented:
Hi, sorry for the delay, but in the meantime `core::error::Error` got stabilized so I removed the `std` feature again. This now requires Rust 1.81.0, if this is too recent I can revert to the previous feature-based solution.

sfackler commented:
I think it'd be safer to have an enabled-by-default `std` feature to avoid being blocked on adding future APIs that do require std. That'd also allow us today to continue to support older Rust versions by importing Error from std when enabled.

linkmauve commented:
Hi, sorry for the very long delay, I’ve added back the std feature to support Rust versions older than 1.81.0, although nowadays that’s a bit old and most of the ecosystem won’t build on such ancient versions any longer.

I’ve added a test in CI for `no_std` on Rust 1.81.0, and a test with the `std` feature on Rust 1.56.0 so that we know there is no regression anywhere.

linkmauve commented:
I’ve also disabled the `std` feature in all of the dependencies, so that it can actually be used in a `no_std` context.
